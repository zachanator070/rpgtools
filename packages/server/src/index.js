import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
