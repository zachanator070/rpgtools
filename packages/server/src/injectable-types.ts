export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
