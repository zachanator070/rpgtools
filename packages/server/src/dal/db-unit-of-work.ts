import {
	ArticleRepository,
	ChunkRepository,
	FileRepository,
	GameRepository,
	ImageRepository,
	ItemRepository,
	ModelRepository,
	MonsterRepository,
	PermissionAssignmentRepository,
	PersonRepository,
	PinRepository,
	PlaceRepository,
	RoleRepository,
	ServerConfigRepository,
	UnitOfWork,
	UserRepository,
	WikiFolderRepository,
	WikiPageRepository,
	WorldRepository,
} from "../types";
import { inject } from "inversify";
import { INJECTABLE_TYPES } from "../injectable-types";

export class DbUnitOfWork implements UnitOfWork {
	@inject(INJECTABLE_TYPES.ArticleRepository)
	articleRepository: ArticleRepository;
	@inject(INJECTABLE_TYPES.ChunkRepository)
	chunkRepository: ChunkRepository;
	@inject(INJECTABLE_TYPES.FileRepository)
	fileRepository: FileRepository;
	@inject(INJECTABLE_TYPES.GameRepository)
	gameRepository: GameRepository;
	@inject(INJECTABLE_TYPES.ImageRepository)
	imageRepository: ImageRepository;
	@inject(INJECTABLE_TYPES.ItemRepository)
	itemRepository: ItemRepository;
	@inject(INJECTABLE_TYPES.ModelRepository)
	modelRepository: ModelRepository;
	@inject(INJECTABLE_TYPES.MonsterRepository)
	monsterRepository: MonsterRepository;
	@inject(INJECTABLE_TYPES.PermissionAssignmentRepository)
	permissionAssignmentRepository: PermissionAssignmentRepository;
	@inject(INJECTABLE_TYPES.PersonRepository)
	personRepository: PersonRepository;
	@inject(INJECTABLE_TYPES.PinRepository)
	pinRepository: PinRepository;
	@inject(INJECTABLE_TYPES.PlaceRepository)
	placeRepository: PlaceRepository;
	@inject(INJECTABLE_TYPES.RoleRepository)
	roleRepository: RoleRepository;
	@inject(INJECTABLE_TYPES.ServerConfigRepository)
	serverConfigRepository: ServerConfigRepository;
	@inject(INJECTABLE_TYPES.UserRepository)
	userRepository: UserRepository;
	@inject(INJECTABLE_TYPES.WikiFolderRepository)
	wikiFolderRepository: WikiFolderRepository;
	@inject(INJECTABLE_TYPES.WikiPageRepository)
	wikiPageRepository: WikiPageRepository;
	@inject(INJECTABLE_TYPES.WorldRepository)
	worldRepository: WorldRepository;

	constructor() {
		// here you would create a generic db context and share with all the repos, the generic db context will share
		// a type of context for each type of db being used
	}
	commit = async (): Promise<void> => {
		return;
	};
}
