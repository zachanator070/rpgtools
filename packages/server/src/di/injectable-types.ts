import {DatabaseContext} from "../dal/database-context";

export const INJECTABLE_TYPES = {

	DomainEntity: Symbol.for("DomainEntity"),

	// domain entities
	Article: Symbol.for("Article"),
	Chunk: Symbol.for("Chunk"),
	File: Symbol.for("File"),
	Game: Symbol.for("Game"),
	Image: Symbol.for("Image"),
	Item: Symbol.for("Item"),
	Model: Symbol.for("Model"),
	Monster: Symbol.for("Monster"),
	PermissionAssignment: Symbol.for("PermissionAssignment"),
	Person: Symbol.for("Person"),
	Pin: Symbol.for("Pin"),
	Place: Symbol.for("Place"),
	Role: Symbol.for("Role"),
	ServerConfig: Symbol.for("ServerConfig"),
	User: Symbol.for("User"),
	WikiFolder: Symbol.for("WikiFolder"),
	World: Symbol.for("World"),
	Event: Symbol.for("Event"),
	Calendar: Symbol.for("Calendar"),

	// entity factories
	ArticleFactory: Symbol.for("ArticleFactory"),
	AclFactory: Symbol.for('AclFactory'),
	ChunkFactory: Symbol.for("ChunkFactory"),
	FileFactory: Symbol.for("FileFactory"),
	GameFactory: Symbol.for("GameFactory"),
	ImageFactory: Symbol.for("ImageFactory"),
	ItemFactory: Symbol.for("ItemFactory"),
	ModelFactory: Symbol.for("ModelFactory"),
	MonsterFactory: Symbol.for("MonsterFactory"),
	PersonFactory: Symbol.for("PersonFactory"),
	PinFactory: Symbol.for("PinFactory"),
	PlaceFactory: Symbol.for("PlaceFactory"),
	RoleFactory: Symbol.for("RoleFactory"),
	ServerConfigFactory: Symbol.for("ServerConfigFactory"),
	UserFactory: Symbol.for("UserFactory"),
	WikiFolderFactory: Symbol.for("WikiFolderFactory"),
	WikiPageFactory: Symbol.for("WikiPageFactory"),
	WorldFactory: Symbol.for("WorldFactory"),
	EventWikiFactory: Symbol.for("EventFactory"),
	CalendarFactory: Symbol.for("CalendarFactory"),

	CharacterFactory: Symbol.for('CharacterFactory'),
	CharacterAttributeFactory: Symbol.for('CharacterAttributeFactory'),
	PathNodeFactory: Symbol.for('NodePathFactory'),
	StrokeFactory: Symbol.for('StrokeFactory'),
	FogStrokeFactory: Symbol.for('FogStrokeFactory'),
	MessageFactory: Symbol.for('MessageFactory'),
	InGameModelFactory: Symbol.for('InGameModelFactory'),

	AgeFactory: Symbol.for("AgeFactory"),
	MonthFactory: Symbol.for("MonthFactory"),
	DayOfTheWeekFactory: Symbol.for("DayOfTheWeekFactory"),

	// repositories
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	FileRepository: Symbol.for("FileRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	EventWikiRepository: Symbol.for("EventWikiRepository"),
	CalendarRepository: Symbol.for("CalendarRepository"),
	StrokeRepository: Symbol.for('StrokeRepository'),
	FogStrokeRepository: Symbol.for('FogStrokeRepository'),

	SqlPermissionControlledRepository: Symbol.for('SqlPermissionControlledRepository'),

	// authorization policies
	ArticleAuthorizationPolicy: Symbol.for("ArticleAuthorizationPolicy"),
	ChunkAuthorizationPolicy: Symbol.for("ChunkAuthorizationPolicy"),
	FileAuthorizationPolicy: Symbol.for("FileAuthorizationPolicy"),
	GameAuthorizationPolicy: Symbol.for("GameAuthorizationPolicy"),
	ImageAuthorizationPolicy: Symbol.for("ImageAuthorizationPolicy"),
	ItemAuthorizationPolicy: Symbol.for("ItemAuthorizationPolicy"),
	ModelAuthorizationPolicy: Symbol.for("ModelAuthorizationPolicy"),
	MonsterAuthorizationPolicy: Symbol.for("MonsterAuthorizationPolicy"),
	PersonAuthorizationPolicy: Symbol.for("PersonAuthorizationPolicy"),
	PinAuthorizationPolicy: Symbol.for("PinAuthorizationPolicy"),
	PlaceAuthorizationPolicy: Symbol.for("PlaceAuthorizationPolicy"),
	RoleAuthorizationPolicy: Symbol.for("RoleAuthorizationPolicy"),
	ServerConfigAuthorizationPolicy: Symbol.for("ServerConfigAuthorizationPolicy"),
	UserAuthorizationPolicy: Symbol.for("UserAuthorizationPolicy"),
	WikiFolderAuthorizationPolicy: Symbol.for("WikiFolderAuthorizationPolicy"),
	WikiPageAuthorizationPolicy: Symbol.for("WikiPageAuthorizationPolicy"),
	WorldAuthorizationPolicy: Symbol.for("WorldAuthorizationPolicy"),
	CalendarAuthorizationPolicy: Symbol.for("CalendarAuthorizationPolicy"),
	FogStrokeAuthorizationPolicy: Symbol.for('FogStrokeAuthorizationPolicy'),
	StrokeAuthorizationPolicy: Symbol.for('StrokeAuthorizationPolicy'),

	// archive repositories
	ArchiveArticleRepository: Symbol.for("ArchiveArticleRepository"),
	ArchiveChunkRepository: Symbol.for("ArchiveChunkRepository"),
	ArchiveGameRepository: Symbol.for("ArchiveGameRepository"),
	ArchiveImageRepository: Symbol.for("ArchiveImageRepository"),
	ArchiveItemRepository: Symbol.for("ArchiveItemRepository"),
	ArchiveModelRepository: Symbol.for("ArchiveModelRepository"),
	ArchiveMonsterRepository: Symbol.for("ArchiveMonsterRepository"),
	ArchivePermissionAssignmentRepository: Symbol.for("ArchivePermissionAssignmentRepository"),
	ArchivePersonRepository: Symbol.for("ArchivePersonRepository"),
	ArchivePinRepository: Symbol.for("ArchivePinRepository"),
	ArchivePlaceRepository: Symbol.for("ArchivePlaceRepository"),
	ArchiveRoleRepository: Symbol.for("ArchiveRoleRepository"),
	ArchiveServerConfigRepository: Symbol.for("ArchiveServerConfigRepository"),
	ArchiveUserRepository: Symbol.for("ArchiveUserRepository"),
	ArchiveWikiFolderRepository: Symbol.for("ArchiveWikiFolderRepository"),
	ArchiveWikiPageRepository: Symbol.for("ArchiveWikiPageRepository"),
	ArchiveWorldRepository: Symbol.for("ArchiveWorldRepository"),
	ArchiveFileRepository: Symbol.for("ArchiveFileRepository"),

	// services
	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	RoleService: Symbol.for("RoleService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	// session contexts
	SessionContextFactory: Symbol.for("SessionContextFactory"),
	SecurityContextFactory: Symbol.for("SecurityContextFactory"),

	// 5e
	Open5eApiClient: Symbol.for("Open5eApiClient"),
	Dnd5eApiClient: Symbol.for("Dnd5eApiClient"),

	// delta
	DeltaFactory: Symbol.for("DeltaFactory"),

	// cache
	Cache: Symbol.for("Cache"),

	// archive
	ArchiveFactory: Symbol.for("ArchiveFactory"),

	ZipArchiveFactory: Symbol.for("ZipArchiveFactory"),
	ZipArchive: Symbol.for('ZipArchive'),

	// event publisher
	EventPublisher: Symbol.for("EventPublisher"),

	// server
	ApiServer: Symbol.for("ApiServer"),
	ServerProperties: Symbol.for("ServerProperties"),
	RpgToolsServer: Symbol.for('RpgToolsServer'),
	DbEngine: Symbol.for("DbEngine"),
	SqlDbEngine: Symbol.for('SqlDbEngine'),

	// data loaders
	ArticleDataLoader: Symbol.for("ArticleDataLoader"),
	ChunkDataLoader: Symbol.for("ChunkDataLoader"),
	FileDataLoader: Symbol.for("FileDataLoader"),
	GameDataLoader: Symbol.for("GameDataLoader"),
	ImageDataLoader: Symbol.for("ImageDataLoader"),
	ItemDataLoader: Symbol.for("ItemDataLoader"),
	ModelDataLoader: Symbol.for("ModelDataLoader"),
	MonsterDataLoader: Symbol.for("MonsterDataLoader"),
	PermissionAssignmentDataLoader: Symbol.for("PermissionAssignmentDataLoader"),
	PersonDataLoader: Symbol.for("PersonDataLoader"),
	PinDataLoader: Symbol.for("PinDataLoader"),
	PlaceDataLoader: Symbol.for("PlaceDataLoader"),
	RoleDataLoader: Symbol.for("RoleDataLoader"),
	ServerConfigDataLoader: Symbol.for("ServerConfigDataLoader"),
	UserDataLoader: Symbol.for("UserDataLoader"),
	WikiFolderDataLoader: Symbol.for("WikiFolderDataLoader"),
	WikiPageDataLoader: Symbol.for("WikiPageDataLoader"),
	WorldDataLoader: Symbol.for("WorldDataLoader"),
	CalendarDataLoader: Symbol.for('CalendarDataLoader'),
	EventDataLoader: Symbol.for('EventDataLoader'),

	// seeders
	RoleSeeder: Symbol.for("RoleSeeder"),
	ServerConfigSeeder: Symbol.for("ServerConfigSeeder"),

	// database context
	DatabaseContext: Symbol.for("DatabaseContext"),
	DatabaseContextFactory: Symbol.for("DatabaseContextFactory"),

	// repo mapper
	EntityMapper: Symbol.for("EntityMapper"),

	// mongodb
	FilterFactory: Symbol.for("FilterFactory"),
	MongoDbMigrationV40: Symbol.for("MongoDbMigrationV40"),


};
