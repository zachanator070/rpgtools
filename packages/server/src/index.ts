import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
