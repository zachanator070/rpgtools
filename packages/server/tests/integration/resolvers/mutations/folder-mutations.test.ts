import { WIKI_FOLDER } from "@rpgtools/common/src/type-constants";
import { WIKI_RW } from "@rpgtools/common/src/permission-constants";
import { container } from "../../../../src/di/inversify";
import { INJECTABLE_TYPES } from "../../../../src/di/injectable-types";
import {CREATE_FOLDER, DELETE_FOLDER, RENAME_FOLDER} from "@rpgtools/common/src/gql-mutations";
import {AuthorizationService} from "../../../../src/services/authorization-service";
import {WikiPageService} from "../../../../src/services/wiki-page-service";
import {Factory} from "../../../../src/types";
import {DbUnitOfWork} from "../../../../src/dal/db-unit-of-work";
import {TEST_INJECTABLE_TYPES} from "../../injectable-types";
import {DefaultTestingContext} from "../../default-testing-context";

process.env.TEST_SUITE = "folder-mutations-test";

describe("folder-mutations", () => {

	const unitOfWorkFactory = container.get<Factory<DbUnitOfWork>>(INJECTABLE_TYPES.DbUnitOfWorkFactory);
	const testingContext = container.get<DefaultTestingContext>(TEST_INJECTABLE_TYPES.DefaultTestingContext);

	const authorizationService = container.get<AuthorizationService>(
		INJECTABLE_TYPES.AuthorizationService
	);
	const wikiPageService = container.get<WikiPageService>(INJECTABLE_TYPES.WikiPageService);

	describe("with tester logged in and world created", () => {
		beforeEach(async () => {
			await testingContext.reset();
		});

		test("create folder", async () => {
			const result = await testingContext.server.executeGraphQLQuery({
				query: CREATE_FOLDER,
				variables: {
					name: "new folder",
					parentFolderId: testingContext.world.rootFolder.toString(),
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					createFolder: {
						_id: expect.any(String),
					},
				},
			});
		});

		test("rename folder", async () => {
			const result = await testingContext.server.executeGraphQLQuery({
				query: RENAME_FOLDER,
				variables: {
					name: "new folder",
					folderId: testingContext.world.rootFolder.toString(),
				},
			});
			expect(result).toMatchSnapshot({
				data: {
					renameFolder: {
						_id: expect.any(String),
					},
				},
			});
		});

		test("delete folder", async () => {
			const result = await testingContext.server.executeGraphQLQuery({
				query: DELETE_FOLDER,
				variables: { folderId: testingContext.newFolder._id.toString() },
			});
			expect(result).toMatchSnapshot({
				data: {
					deleteFolder: {
						_id: expect.any(String),
					},
				},
			});
		});

		test("delete root folder", async () => {
			const result = await testingContext.server.executeGraphQLQuery({
				query: DELETE_FOLDER,
				variables: { folderId: testingContext.world.rootFolder.toString() },
			});
			expect(result).toMatchSnapshot();
		});

		describe("with unauthenticated user", () => {
			beforeEach(async () => {
				testingContext.mockSessionContextFactory.resetCurrentUser();
			});

			test("create folder permission denied", async () => {
				const result = await testingContext.server.executeGraphQLQuery({
					query: CREATE_FOLDER,
					variables: {
						name: "new folder",
						parentFolderId: testingContext.world.rootFolder.toString(),
					},
				});
				expect(result).toMatchSnapshot();
			});

			test("rename folder permission denied", async () => {
				const result = await testingContext.server.executeGraphQLQuery({
					query: RENAME_FOLDER,
					variables: {
						name: "new folder",
						folderId: testingContext.world.rootFolder.toString(),
					},
				});
				expect(result).toMatchSnapshot();
			});

			test("delete folder permission denied", async () => {
				const result = await testingContext.server.executeGraphQLQuery({
					query: DELETE_FOLDER,
					variables: { folderId: testingContext.newFolder._id.toString() },
				});
				expect(result).toMatchSnapshot({
					errors: [
						{
							message: expect.any(String),
						},
					],
				});
			});
		});

		describe("as other user", () => {
			beforeEach(async () => {
				await testingContext.reset();
				testingContext.mockSessionContextFactory.setCurrentUser(testingContext.otherUser);
				const unitOfWork = unitOfWorkFactory({});
				await wikiPageService.createWiki(testingContext.testerSecurityContext, "new page", testingContext.newFolder._id, unitOfWork);
				await authorizationService.grantUserPermission(
					testingContext.testerSecurityContext,
					WIKI_RW,
					testingContext.newFolder._id,
					WIKI_FOLDER,
					testingContext.otherUser._id,
					unitOfWork
				);
				await unitOfWork.commit();
				testingContext.mockSessionContextFactory.setCurrentUser(testingContext.otherUser);
			});

			test("no permission to child", async () => {
				const result = await testingContext.server.executeGraphQLQuery({
					query: DELETE_FOLDER,
					variables: { folderId: testingContext.newFolder._id.toString() },
				});
				expect(result).toMatchSnapshot({
					errors: [
						{
							message: expect.any(String),
						},
					],
				});
			});
		});
	});
});
