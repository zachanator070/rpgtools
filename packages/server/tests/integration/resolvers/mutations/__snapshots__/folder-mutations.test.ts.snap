// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`folder-mutations with tester logged in and world created as other user no permission to child 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": Any<String>,
      "path": Array [
        "deleteFolder",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created create folder 1`] = `
Object {
  "data": Object {
    "createFolder": Object {
      "_id": Any<String>,
      "children": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "name": Any<String>,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created delete folder 1`] = `
Object {
  "data": Object {
    "deleteFolder": ObjectContaining {
      "_id": Any<String>,
      "children": ArrayContaining [
        ObjectContaining {
          "_id": Any<String>,
          "name": Any<String>,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created delete root folder 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You cannot delete the root folder of a world],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created rename folder 1`] = `
Object {
  "data": Object {
    "renameFolder": Object {
      "_id": Any<String>,
      "name": "new folder",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created with unauthenticated user create folder permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission for this folder],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created with unauthenticated user delete folder permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": Any<String>,
      "path": Array [
        "deleteFolder",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`folder-mutations with tester logged in and world created with unauthenticated user rename folder permission denied 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You do not have permission for this folder],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
